module/apmgopg/hook.go

```diff
@@ -0,0 +1,72 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// +build go1.9
+
+package apmgopg
+
+import (
+	"github.com/go-pg/pg"
+
+	"go.elastic.co/apm"
+	"go.elastic.co/apm/module/apmsql"
+)
+
+const elasticApmSpanKey = "go-apm-agent:span"
+
+type QueryHook struct {
+	Database string
```

Comment:
Is it necessary to store Database and Username here? Can the hook use https://godoc.org/github.com/go-pg/pg#DB.Options to get them instead? I _think_ that the `orm.DB` field is actually the `*pg.DB` value, so you could do a type-assertion.

Reply:
I just tried this, and it didn't work *pb.DB doesn't implement orm.DB. :/

Reply:
I just tried this locally and it worked:

```diff
diff --git a/module/apmgopg/hook.go b/module/apmgopg/hook.go                                         
index fa6a1ece..785e39ba 100644                                                                      
--- a/module/apmgopg/hook.go                                                                         
+++ b/module/apmgopg/hook.go                                                                         
@@ -28,27 +28,35 @@ import (                                                                         
 const elasticApmSpanKey = "go-apm-agent:span"                                                       
                                                                                                     
 // QueryHook is an implementation of pg.QueryHook that reports queries as spans to Elastic APM.     
-type QueryHook struct {                                                                             
-       Database string                                                                              
-       Username string                                                                              
-}                                                                                                   
+type QueryHook struct{}                                                                             
                                                                                                     
 // BeforeQuery initiates the span for the database query                                            
 func (qh *QueryHook) BeforeQuery(evt *pg.QueryEvent) {                                              
+       var (                                                                                        
+               database string                                                                      
+               user     string                                                                      
+       )                                                                                            
+       db, ok := evt.DB.(*pg.DB)                                                                    
+       if ok {                                                                                      
+               opts := db.Options()                                                                 
+               database = opts.Database                                                             
+               user = opts.User                                                                     
+       }                                                                                            
+                                                                                                    
        sql, err := evt.UnformattedQuery()                                                           
        if err != nil {                                                                              
                // Expose the error making it a bit easier to debug                                  
                sql = fmt.Sprintf("[go-pg] error: %s", err.Error())                                  
        }                                                                                            
                                                                                                     
-       span, _ := apm.StartSpan(evt.DB.Context(), apmsql.QuerySignarture(sql), "db.postgres.query") 
+       span, _ := apm.StartSpan(evt.DB.Context(), apmsql.QuerySignature(sql), "db.postgresql.query")
        span.Context.SetDatabase(apm.DatabaseSpanContext{                                            
                Statement: sql,                                                                      
                                                                                                     
                // Static                                                                            
                Type:     "sql",                                                                     
-               User:     qh.Username,                                                               
-               Instance: qh.Database,                                                               
+               User:     user,                                                                      
+               Instance: database,                                                                  
        })                                                                                           
                                                                                                     
        evt.Data[elasticApmSpanKey] = span                                                           
```